notebooks/04_model_select_and_optimize/nni_ncf.ipynb

```diff
@@ -0,0 +1,762 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "stop_nni()\n",
+    "config_path_ncf = os.path.join(TMP_DIR, 'config_ncf.yml')\n",
+    "with Timer() as time_ncf:\n",
+    "    start_nni(config_path_ncf, wait=WAITING_TIME, max_retries=MAX_RETRIES)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "check_metrics_written(wait=WAITING_TIME, max_retries=MAX_RETRIES)\n",
+    "trials_ncf, best_metrics_ncf, best_params_ncf, best_trial_path_ncf = get_trials('maximize')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'rmse': 3.2212178182820117,\n",
+       " 'ndcg_at_k': 0.1439899349063176,\n",
+       " 'precision_at_k': 0.11633085896076353}"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "best_metrics_ncf"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{'parameter_id': 3,\n",
+       " 'parameter_source': 'algorithm',\n",
+       " 'parameters': {'n_factors': 12, 'learning_rate': 0.0023365527461525885},\n",
+       " 'parameter_index': 0}"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "best_params_ncf"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Baseline Model\n",
+    "\n",
+    "Although we hope that the additional effort of utilizing an AutoML framework like NNI for hyperparameter tuning will lead to better results, we should also draw comparisons using our baseline model (our model trained with its default hyperparameters). This allows us to precisely understand what performance benefits NNI is or isn't providing."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data = NCFDataset(train, validation, seed=DEFAULT_SEED)\n",
+    "model = NCF(\n",
+    "    n_users=data.n_users, \n",
+    "    n_items=data.n_items,\n",
+    "    model_type=\"NeuMF\",\n",
+    "    n_factors=4,\n",
+    "    layer_sizes=[16,8,4],\n",
+    "    n_epochs=NUM_EPOCHS,\n",
+    "    learning_rate=1e-3,  \n",
+    "    verbose=True,\n",
+    "    seed=DEFAULT_SEED\n",
+    ")\n",
+    "model.fit(data)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def compute_test_results(model, train, test):\n",
```

Comment:
i think functions belong in the code section, and should just be used in notebooks. Will keep the users focused on the content within the notebook.

Reply:
refactored